pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/FilterOperator.java

```diff
@@ -96,20 +96,21 @@ public String toExplainString() {
   @Override
   protected TransferableBlock getNextBlock() {
     TransferableBlock block = _input.nextBlock();
-    if (block.isEndOfStreamBlock()) {
-      if (block.isErrorBlock()) {
-        return block;
-      }
+    if (block.isErrorBlock()) {
+      return block;
+    }
+    if (block.isSuccessfulEndOfStreamBlock()) {
       return updateEosBlock(block, _statMap);
     }
-    List<Object[]> resultRows = new ArrayList<>();
+    assert block.isDataBlock();
+    List<Object[]> rows = new ArrayList<>();
     for (Object[] row : block.getContainer()) {
       Object filterResult = _filterOperand.apply(row);
       if (BooleanUtils.isTrueInternalValue(filterResult)) {
-        resultRows.add(row);
+        rows.add(row);
       }
     }
-    return new TransferableBlock(resultRows, _dataSchema, DataBlock.Type.ROW);
+    return rows.isEmpty() ? getNextBlock() : new TransferableBlock(rows, _dataSchema, DataBlock.Type.ROW);
```

Comment:
This may be a bit difficult to read, but I guess the idea is that next block will always be a EOS, right?

Reply:
No. Basically we want to skip the empty block, until we find an non-empty block, or finished all data. Noted that we can get empty block when all rows in upstream block is filtered out.
Let me make it a while loop which is probably easier to understand